pkg/blockbuilder/scheduler/scheduler.go

```
@@ -2,43 +2,134 @@ package scheduler
+	planner := NewTimeRangePlanner(cfg.TargetRecordConsumptionPeriod, offsetReader, func() time.Time { return time.Now().UTC() }, logger)
+	s := &BlockScheduler{
+		cfg:          cfg,
+		planner:      planner,
+		offsetReader: offsetReader,
+		logger:       logger,
+		metrics:      NewMetrics(r),
+		queue:        queue,
+	}
+	s.Service = services.NewBasicService(nil, s.running, nil)
+	return s
+}
+
+func (s *BlockScheduler) running(ctx context.Context) error {
+	ticker := time.NewTicker(s.cfg.Interval)
+	for {
+		select {
+		case <-ticker.C:
+			if err := s.runOnce(ctx); err != nil {
+				// TODO: add metrics
+				level.Error(s.logger).Log("msg", "failed to schedule jobs", "err", err)
+			}
+		case <-ctx.Done():
+			return nil
+		}
 	}
 }
 
-func (s *QueueScheduler) HandleGetJob(ctx context.Context, builderID string) (*types.Job, bool, error) {
+func (s *BlockScheduler) runOnce(ctx context.Context) error {
+	lag, err := s.offsetReader.GroupLag(ctx)
+	if err != nil {
+		level.Error(s.logger).Log("msg", "failed to get group lag", "err", err)
+		return err
+	}
+
+	if err := s.publishLagMetrics(lag); err != nil {
+		level.Error(s.logger).Log("msg", "failed to publish lag metrics", "err", err)
+	}
+
+	jobs, err := s.planner.Plan(ctx)
+	if err != nil {
+		level.Error(s.logger).Log("msg", "failed to plan jobs", "err", err)
+	}
+
+	for _, job := range jobs {
+		// TODO: end offset keeps moving each time we plan jobs, maybe we should not use it as part of the job ID
+		if status, ok := s.queue.Exists(&job); ok {
+			level.Debug(s.logger).Log("msg", "job already exists", "job", job, "status", status)
+			continue
+		}
+
+		if err := s.queue.Enqueue(&job); err != nil {
+			level.Error(s.logger).Log("msg", "failed to enqueue job", "job", job, "err", err)
+		}
+	}
+
+	return nil
+}
+
+func (s *BlockScheduler) publishLagMetrics(lag map[int32]kadm.GroupMemberLag) error {
```

don't think this should return an error.